src/interactive.rs

```diff
@@ -0,0 +1,209 @@
+use async_std::channel::{unbounded, Receiver, Sender};
+use async_std::sync::{Arc, Mutex};
+use async_std::{io};
+
+use crate::ast::{Rule, Term};
+use crate::parser::parse_terms;
+use crate::prolog::codegen::{generate, generate_query};
+use crate::prolog::util::AssertLocation;
+use crate::prolog::{pred, ActivatedEngine, Context, ContextExt, Module};
+use crate::view::{State, View};
+
+pub struct PrologHandler<'a> {
+  ctx:Context<'a, ActivatedEngine<'a>>,
+  sender:Sender<String>,
+  receiver:Receiver<Vec<Term>>,
+}
+
+impl<'a> PrologHandler<'a> {
+  pub async fn handle(&mut self) {
+    if let Ok(terms) = self.receiver.recv().await {
+      if let Ok((list, other)) = generate_query(&self.ctx, &terms) {
+        if let Ok(()) = self.ctx.call_once(pred![mlatu: rewrite / 2], [&list, &other]) {
```

Comment:
(Can't access discord right now)
This is probably the issue; generate_query, ctx.call_once, and ctx.canonical all do stuff with the prolog FFI, which is not async. This messes up the async executor (slowing things down more than it would already), so you need to spawn a blocking task to handle this. I can try to work on a fix when I get home, if you want me to

Reply:
That would be great if you are able and have an idea of the problem. I can keep working on this as well and see if there’s a way to spawn a thread which uses a non-Send resource